### PR TITLE
PWA-1841 & PWA-2302: Convert CSS to Tailwind

### PR DESCRIPTION
**Pull request by jimbo**
## Description

Convert as many CSS rules to Tailwind (via `composes`) as possible.

- Originally authored by @brendanfalkowski 
- Source branch: https://github.com/brendanfalkowski/pwa-studio/commits/brendan/PWA-1895-monarch

## Related Issue

Closes PWA-1841 & PWA-2302.

### Verification Stakeholders

- @brendanfalkowski 
- @jcalcaben 

#### Test scenario(s) for direct fix/feature

Full regression needed; some discrepancies are acceptable within reason.

#### Test scenario(s) for any Magento Backend Supported Configurations

--

#### Is Browser/Device testing needed?

Yes; all components are affected across all devices.

#### Any ad-hoc/edge case scenarios that need to be considered?

TBD.

## Breaking Changes (if any)

Components TBD. All styles are changing.

## Checklist

-   I have added tests to cover my changes, if necessary.
-   I have added translations for new strings, if necessary.
-   I have updated the documentation accordingly, if necessary.

### CODE REVIEW

**Comment by brendanfalkowski** [comment]
@jimbo Heads up, Prettier broke all the arbitrary values in Tailwind that contain spaces. TW specifies these have to be written without spaces to work. Example:

`checkoutPage.module.css`

```
# Original
composes: lg_grid-cols-[2fr,1fr] from global;

# Prettier version, space added, TW won't generate the rule so it doesn't apply
composes: lg_grid-cols-[2fr, 1fr] from global;
```

Screen Shot 2022-02-03 at 12 11 51 PM

**Comment by fooman** [comment]
Might an [underscore](https://tailwindcss.com/docs/adding-custom-styles#handling-whitespace) work? 

```
<div class="grid grid-cols-[1fr_500px_2fr]">
  <!-- ... -->
</div>
```

**Comment by brendanfalkowski** [comment]
@fooman That should solve the conflict Prettier causes after we update to TW3. This build is still on TW2 because the work started before TW3 was released. It's probably why they changed from `,` to `_` for a space-replacement character in arbitrary values.

Under TW2 using `_` will generate the class, but the character isn't replaced, so the final CSS isn't valid.

Screen Shot 2022-02-14 at 9 22 24 AM

Temporarily, we'll have to wrap the broken declarations with `prettier ignore` comments so we don't lose track of the regressions, and revisit in the TW3 update.

File: packages/pwa-theme-venia/package.json [hunk 1]
```diff
@@ -1,10 +1,10 @@
 {
-  "name": "@magento/pwa-theme-venia",
+  "name": "@magento/pwa-theme-base",
```
Comment (jcalcaben) [inline packages/pwa-theme-venia/package.json:2]: Should we hold off renaming the package until the next release? Technically, this is still the Venia theme.

**Comment by drpayyne** [comment]
What's the reason behind composing Tailwind classes from the global namespace instead of embedding them directly in HTML tags? From first glance, it looks like there's a lot of repetition so I'm curious as to what's the reason causing this to be less-important. Thanks.

**Comment by brendanfalkowski** [comment]
@drpayyne Using Tailwind inside CSS Modules isn't a perfect solution (it causes some specific issues), but it solves most of the problems this solution was architected for: support theming without changing JSX or requiring TW adoption for backward compatibility.

There is a lot of repetition when authoring, but it doesn't affect the build size so the benefits of Tailwind aren't lost. The classes written into JSX components are essentially a static API. Existing implementations aren't required to use Tailwind, and if all JSX changed we'd be forcing them to migrate all their custom CSS into TW as most CSS imports would be removed. Additionally, extensions can use the "targets" functionality on CSS without having affecting JSX, which reduces the long-term maintenance scope for implementations.

File: packages/venia-ui/lib/components/ToastContainer/toast.module.css [hunk 2]
```diff
@@ -106,17 +113,19 @@
 }
 
 .controls {
+    composes: border from global;
```
Comment (jcalcaben) [inline packages/venia-ui/lib/components/ToastContainer/toast.module.css:116]: Should be `border-l`

**Review by jcalcaben — CHANGES_REQUESTED** [review]
Some initial issues I noticed while clicking through the storefront and comparing

**Review by jcalcaben — CHANGES_REQUESTED** [review]
Additional feedback

File: packages/venia-ui/lib/components/clickable.module.css [hunk 3]
```diff
@@ -1,9 +1,9 @@
 .root {
-    align-items: center;
-    cursor: pointer;
-    display: inline-flex;
-    justify-content: center;
-    line-height: 1;
-    pointer-events: auto;
-    text-align: center;
+    composes: items-center from global;
```
[the remark this answers is not in the document]
Comment (jimbo) [reply packages/venia-ui/lib/components/clickable.module.css]: Done.

File: packages/venia-ui/lib/components/Button/button.module.css [hunk 4]
```diff
@@ -1,94 +1,114 @@
...
-.root:hover {
-    --stroke: var(--venia-brand-color-1-800);
-}
+    composes: active_border-brand-darkest from global;
+    composes: active_text-brand-darkest from global;
 
-.root:focus {
-    box-shadow: -6px 6px rgb(var(--venia-brand-color-1-100));
+    composes: disabled_border-gray-400 from global;
+    composes: disabled_pointer-events-none from global;
```
[the remark this answers is not in the document]
Comment (jimbo) [reply packages/venia-ui/lib/components/Button/button.module.css:31]: Fixed.

File: packages/pwa-theme-venia/tailwind.preset.js [hunk 5]
```diff
@@ -1,39 +1,255 @@
...
+        )}, white)`,
+        swatch: theme('colors.gray.100'),
+        'swatch-selected': `linear-gradient(-45deg, rgba(0, 0, 0, 0.2), transparent), ${theme(
+            'colors.gray.100'
+        )}`
+    }),
+    borderColor: theme => ({
+        button: theme('colors.gray.600'),
+        error: theme('colors.red.400'),
+        info: theme('colors.emerald.600'),
```
[the remark this answers is not in the document]
Comment (jimbo) [reply packages/pwa-theme-venia/tailwind.preset.js]: Yeah, they fixed this in Tailwind 3, which we aren't using yet. Fixed.

File: packages/venia-ui/lib/components/ToastContainer/toast.module.css (hunk 2 above)
Comment (jimbo) [reply packages/venia-ui/lib/components/ToastContainer/toast.module.css]: Fixed.

File: packages/venia-ui/lib/components/AccountMenu/accountMenuItems.module.css [hunk 6]
```diff
@@ -1,26 +1,26 @@
...
-    padding: 1.5rem;
+    composes: gap-y-sm from global;
+    composes: grid from global;
+    composes: grid-flow-row from global;
+    composes: justify-stretch from global;
+    composes: p-sm from global;
 }
 
 .link {
     composes: root from '../LinkButton/linkButton.module.css';
```
[the remark this answers is not in the document]
Comment (jimbo) [reply packages/venia-ui/lib/components/AccountMenu/accountMenuItems.module.css]: Agree. We should get rid of all cross-component dependencies, since Tailwind handles deduplication for us now automatically. If we really need some shared CSS, we can write a Tailwind plugin.

File: packages/venia-ui/lib/components/AddressBookPage/addressBookPage.module.css [hunk 7]
```diff
@@ -1,56 +1,45 @@
...
 }
 
-@media (max-width: 1024px) {
-    .root {
-        padding-left: 1.5rem;
-        padding-right: 1.5rem;
-    }
+.addButton {
+    composes: border-2 from global;
+    composes: border-solid from global;
```
[the remark this answers is not in the document]
Comment (jimbo) [reply packages/venia-ui/lib/components/AddressBookPage/addressBookPage.module.css]: Fixed.

File: packages/venia-ui/lib/components/LegacyMiniCart/kebab.module.css [hunk 8]
```diff
@@ -1,46 +1,37 @@
...
+    composes: mt-1 from global;
+    composes: relative from global;
+    composes: row-end-4 from global;
+    composes: row-start-1 from global;
 }
 
 .dropdown {
-    align-items: center;
+    composes: absolute from global;
+    composes: bg-xxx from global;
```
[the remark this answers is not in the document]
Comment (jimbo) [reply packages/venia-ui/lib/components/LegacyMiniCart/kebab.module.css]: No. Removed this.

File: packages/pwa-theme-venia/tailwind.preset.js [hunk 9]
```diff
@@ -1,39 +1,255 @@
...
+    },
+    zIndex: {
+        behind: '-1',
+        surface: '1',
+        foreground: '10',
+        header: '15',
+        button: '20',
+        buttonHover: '21',
+        buttonFocus: '22',
+        dropdown: '50',
```
[the remark this answers is not in the document]
Comment (jimbo) [reply packages/pwa-theme-venia/tailwind.preset.js]: Fixed. Let me know if you spot any issues as a result of my change here—seems good to me.

File: packages/venia-ui/lib/components/LegacyMiniCart/kebab.module.css (hunk 8 above)
[the remark this answers is not in the document]
Comment (jcalcaben) [reply packages/venia-ui/lib/components/LegacyMiniCart/kebab.module.css]: I forgot to mention this entry is also missing `composes: bg-white from global;`
Right now it's rendering a clear background.

**Comment by jimbo** [comment]
Merged with `develop` and fixed a small regression.

**Comment by anthoula** [comment]
✅  Small regression fix looks good to me. James C approved prior changes. Ready for QA

**Review by anthoula — DISMISSED** [review]

**Comment by dpatil-magento** [comment]
run cypress

**Comment by dpatil-magento** [comment]
@jimbo Adding Observations part manual UI testing -

1. Cart page shipping methods displaying in row horizontally instead of vertically.  (Exists in desktop as well)
2. Edit overlays like Cart > Product Edit, Checkout > Shipping, Payment  >>>>> Displaying only half UI. (Exists only in Mobile)
3. Scrolling on both Edit. overlay or Page does not scroll to top.  Example - In mobile view click on Review Order button OR on checkout page scroll to shipping section and hit edit icon. (Exists only in Mobile)
4. Order Confirmation UI for guest is broken. (Exists only in Mobile)
5. Left side main menu overlay >> Store switcher and account sign In are not visible by default …user has to scroll.   On develop they are visible by default. Same issue with Filter overlay >> See Results   must be visible by default. but user has to scroll to reach to it. (Exists only in Mobile)
6. My account. >Communications >>>>  No vertical space  between Venia header Row and Communications  page title. Same issue with My account >> Address book and Saved Payments pages.
7. For auth user on checkout. page hit Edit for shipping overlay >> User gets blank overlay.  (Exists in desktop as well)
8. If more than 2 products. in cart then checkout page product container is broken.  (Exists in desktop as well)
9. DELETE  button on Auth > Shipping > Delete Shipping is not Red color.  (Exists in desktop as well) (edited) 
10. New commits broke something major. Check page width wrt content.
11. Mobile view Newsletter textfiles width breaking footer area.

**Comment by dpatil-magento** [comment]
run cypress

**Comment by dpatil-magento** [comment]
run cypress

**Comment by dpatil-magento** [comment]
run cypress

**Comment by dpatil-magento** [comment]
run cypress

**Comment by dpatil-magento** [comment]
run cypress

**Comment by dpatil-magento** [comment]
run lighthouse-desktop

**Comment by dpatil-magento** [comment]
run lighthouse-mobile

**Comment by dpatil-magento** [comment]
run lighthouse-mobile

**Comment by dpatil-magento** [comment]
run lighthouse-mobile

**Comment by dpatil-magento** [comment]
QA Approved.

**Review by anthoula — APPROVED** [review]
Code Review LGTM ⭐

**Comment by brendanfalkowski** [comment]
🎉🎉🎉

**Comment by slide13** [comment]
@jimbo I apologize for writing in a closed pr, but tell me, why do we use composes from the global space to write css through css modules with tailwind, except for writing styles through the @apply utility? in this case, we lose autocomplete, and incorrect classes are ignored and simply not applied without errors.
As an example, why not instead of this:

```
    composes: appearance-none from global;
    composes: border-2 from global;
    composes: border-solid from global;
    composes: flex-textInput from global;
    composes: h-[2.5rem] from global;
    composes: inline-flex from global;
    composes: m-0 from global;
    composes: max-w-full from global;
    composes: rounded-md from global;
    composes: text-colorDefault from global;
    composes: w-full from global;
```

use this:

```
@apply appearance-none
    border-2
    border-solid
    flex-textInput
    h-[2.5rem]
    inline-flex
    m-0
    max-w-full
    rounded-md
    text-colorDefault
    w-full
```
